File: scala/src/main/scala/yuzu/jsonld.scala
package com.github.jmccrae.yuzu

import com.github.jmccrae.yuzu.YuzuSettings._
import com.hp.hpl.jena.rdf.model.{Literal, Model, RDFNode, Resource, Property}
import com.hp.hpl.jena.vocabulary._
import java.io.Writer
import scala.collection.JavaConversions._
import scala.collection.mutable.{Map => MutableMap}

object JsonLDPrettySerializer {
  sealed trait JsonObj {
    def write(out : Writer, indent : Int) : Unit
    override def toString = {
      val sw = new java.io.StringWriter()
      write(sw, 0)
      sw.toString }
  }

  object JsonObj {
    def apply(value : Any) : JsonObj = value match {
      case o : JsonObj => 
        o
      case s : String =>
        JsonString(s)
      case s : Seq[_] =>
        JsonList(s.map(JsonObj(_)))
      case m : Map[_, _] =>
        JsonMap(MutableMap((m.map {
          case (k, v) => k.toString -> JsonObj(v)
        }).toSeq:_*))
      case m : MutableMap[_, _] =>
        JsonMap(m.map {
          case (k, v) => k.toString -> JsonObj(v)
        })
    }
    def apply(vals : (String, Any)*) : JsonMap = JsonMap((vals.map {
      case (k, v) => k -> JsonObj(v)
    }):_*)
  }
  case class JsonString(value : String) extends JsonObj {
    def write(out : Writer, indent : Int) {
      out.write("\"%s\"" format (value.replaceAll("\\\\","\\\\\\\\").
        replaceAll("\"", "\\\\\""))) }
  }

  class JsonMap(val value : MutableMap[String, JsonObj]) extends JsonObj {
    def write(out : Writer, indent : Int) {
      out.write("{\n")
      val elems = value.toSeq.sortBy(_._1)
      if(!elems.isEmpty) {
        val h = elems.head
        out.write("  " * (indent + 1))
        out.write("\"%s\": " format h._1)
        h._2.write(out, indent + 2)
        for((k, v) <- elems.tail) {
          out.write(",\n")
          out.write("  " * (indent + 1))
          out.write("\"%s\": " format k)
          v.write(out, indent + 2) }}
      out.write("\n")
      out.write("  " * indent)
      out.write("}") }

    def update(key : String, v : JsonObj) = value.update(key, v)
    def update(key : String, v : String) = value.update(key, JsonString(v))
    def contains(key : String) = value.contains(key)
    def apply(key : String) = value(key)
    def remove(key : String) = value.remove(key)
    def keys = value.keys
    override def equals(o : Any) = o match {
      case null => 
        false
      case jm : JsonMap =>
        value == jm.value
      case _ =>
        false }
  }

  object JsonMap {
    def apply(value : MutableMap[String, JsonObj]) = new JsonMap(value)
    def apply(vals : (String, JsonObj)*) = new JsonMap(
      MutableMap(vals:_*))
  }

  case class JsonList(value : Seq[JsonObj]) extends JsonObj {
    def write(out : Writer, indent : Int) {
      out.write("[\n")
      if(!value.isEmpty) {
        val h = value.head
        out.write("  " * (indent + 1))
        h.write(out, indent + 1)
        for(v <- value.tail) {
          out.write(",\n")
          out.write("  " * (indent + 1))
          v.write(out, indent + 1) }}
      out.write("\n")
      out.write("  " * indent)
      out.write("]") }

    def :+(elem : JsonObj) = JsonList(value :+ elem)
    def +:(elem : JsonObj) = JsonList(elem +: value)
    def apply(i : Int) = value(i)
  }

  implicit class GraphPimp(graph : Model) {
    def listProperties(subj : Resource = null, obj : RDFNode = null) = {
      (graph.listStatements().filter { st =>
        (subj == null || st.getSubject() == subj) &&
        (obj == null || st.getObject() == obj) 
      } map { st =>
        st.getPredicate()
      }).toSet }
  }

  private def isAlnum(s : String) = s.matches("\\w+")

  private def propType(p :  Property, graph : Model) : JsonObj = {
    if(graph.listObjectsOfProperty(p).forall(_.isResource())) {
      JsonObj("@id" -> p.getURI(),
              "@type" -> "@id") }
    else {
      JsonString(p.getURI()) }}

  private def splitURI(value : String) : (Option[String], String) = {
    if(value.startsWith(BASE_NAME)) {
      return (None, value.drop(BASE_NAME.length)) }
    for((uri, qn) <- Seq((PREFIX1_URI, PREFIX1_QN),
                    (PREFIX2_URI, PREFIX2_QN),
                    (PREFIX3_URI, PREFIX3_QN),
                    (PREFIX4_URI, PREFIX4_QN),
                    (PREFIX5_URI, PREFIX5_QN),
                    (PREFIX6_URI, PREFIX6_QN),
                    (PREFIX7_URI, PREFIX7_QN),
                    (PREFIX8_URI, PREFIX8_QN),
                    (PREFIX9_URI, PREFIX9_QN),
                    (RDF.getURI(), "rdf"),
                    (RDFS.getURI(), "rdfs"),
                    (XSD.getURI(), "xsd"),
                    (OWL.getURI(), "owl"),
                    (DC_11.getURI(), "dc"),
                    (DCTerms.getURI(), "dct"))) {
      if(value.startsWith(uri) && uri != "http://www.example.com/") {
        return (Some(qn), value.drop(uri.length)) }}
    return (None, value) }

  private def extractJsonLDContext(graph : Model, 
      query : String) : (JsonMap, Map[String, String]) = {
    val context = JsonObj(
      "@base" -> BASE_NAME,
      PREFIX1_QN -> PREFIX1_URI,
      PREFIX2_QN -> PREFIX2_URI,
      PREFIX3_QN -> PREFIX3_URI,
      PREFIX4_QN -> PREFIX4_URI,
      PREFIX5_QN -> PREFIX5_URI,
      PREFIX6_QN -> PREFIX6_URI,
      PREFIX7_QN -> PREFIX7_URI,
      PREFIX8_QN -> PREFIX8_URI,
      PREFIX9_QN -> PREFIX9_URI,
      "rdf" -> RDF.getURI(),
      "rdfs" -> RDFS.getURI(),
      "xsd" -> XSD.getURI(),
      "owl" -> OWL.getURI(),
      "dc" -> DC_11.getURI(),
      "dct" -> DCTerms.getURI())
    for(k <- context.keys) {
      if(context(k) == JsonString("http://www.example.com/")) {
        context.remove(k) }}
    val props = MutableMap[String, String]()
    for(p <- graph.listProperties()) {
      val pStr = p.getURI()
      var shortName = ""
      if(pStr.contains("#")) {
        shortName = pStr.drop(pStr.lastIndexOf("#") + 1) }
      if(!isAlnum(shortName) && pStr.contains("/")) {
        shortName = pStr.drop(pStr.lastIndexOf("/") + 1) }
      if(!isAlnum(shortName)) {
        val (pre, suf) = splitURI(pStr)
        shortName = suf }
      var sn = shortName
      var i = 2
      while(context.contains(sn)) {
        sn = "%s%d" format (shortName, i)
        i += 1 }
      if(p == RDF.`type`) {
        sn = "@type" }
      else {
        context(sn) = propType(p, graph) }
      props(pStr) = sn }
    (context, props.toMap) } 

  private def addProps(obj : JsonMap, value : Resource, context : JsonMap, graph : Model,
               query : String, prop2sn : Map[String, String], drb : Set[Resource],
               stack : List[Resource]) {
    if(!(stack contains value)) {
      for(p <- graph.listProperties(value)) {
        val objs = graph.listObjectsOfProperty(value, p).toList.sortBy(_.toString)
        val isObj = p == RDF.`type` ||
          context(prop2sn(p.getURI())).isInstanceOf[JsonMap] 
        
        if(objs.size == 1) {
          graph.removeAll(value, p, objs(0))
          obj(prop2sn(p.getURI())) = jsonLDValue(objs(0), context, graph, query,
                                                 prop2sn, isObj, drb, 
                                                 value :: stack) }
        else {
          for(o <- objs) {
            graph.removeAll(value, p, o) }
          obj(prop2sn(p.getURI())) = JsonList((objs.map { o => 
            jsonLDValue(o, context, graph, query, prop2sn, 
                        isObj, drb, value :: stack) 
          }).toList) }}}}


  private def addInverseProps(obj : JsonMap, value : Resource, context : JsonMap, 
                      graph : Model, query : String, 
                      prop2sn : Map[String, String], drb : Set[Resource],
                      stack : List[Resource]) {
    if(!(stack contains value)) {
      for(p <- graph.listProperties()) {
        val objs = graph.listObjectsOfProperty(value, p).toList.sortBy(_.toString)
        for(o <- objs) {
          if(o.isResource()) {
            obj(prop2sn(p.getURI())) match {
              case jm : JsonMap =>
                addInverseProps(jm, value, context, graph, query, prop2sn, drb,
                                value :: stack) 
              case _ =>
                // ignore
      }}}}

      if(!graph.listProperties(null, value).isEmpty) {
        val robj = JsonMap()
        obj("@reverse") = robj
        for(p <- graph.listProperties(null, value)) {
          val objs = graph.listSubjectsWithProperty(p, value).toList.
            sortBy(_.toString).filter { o =>
              o.isResource() //&& !o.asResource().getURI().startsWith(query) 
            } map { o =>
              o.asResource() }

          if(objs.size == 1) {
            graph.removeAll(objs(0), p, value)
            robj(prop2sn(p.getURI())) = jsonLDValue(objs(0), context, graph,
                                                    query, prop2sn, false, drb,
                                                    value :: stack) }
          else if(objs.size > 1) {
            for(o <- objs) {
              graph.removeAll(o, p, value) }
            robj(prop2sn(p.getURI())) = JsonList(objs.map { o =>
              jsonLDValue(o, context, graph, query, prop2sn, 
                          false, drb, value :: stack) 
            }) }}}}}

  private def jsonLDValue(value : Any, context : JsonMap, graph : Model, 
                  query : String, prop2sn : Map[String, String], 
                  isObj : Boolean, drb : Set[Resource],
                  stack : List[Resource]) : JsonObj = {
    value match {
      case list : Seq[_] => 
        if(list.size == 1) {
          jsonLDValue(list(0), context, graph, query, prop2sn, 
                      isObj, drb, stack) }
        else {
          JsonList(list.map { v =>
            jsonLDValue(v, context, graph, query, prop2sn, 
                        isObj, drb, stack) })}
      case r : Resource if r.isURIResource() =>
        if(graph.listStatements(r, null, null : RDFNode).isEmpty && isObj) {
          val (pre, suf) = splitURI(r.getURI())
          pre match {
            case Some(pre) =>
              JsonString("%s:%s" format (pre, suf))
            case None => 
              JsonString(suf) }}
        else {
          val (pre, suf) = splitURI(r.getURI())
          val obj = pre match {
            case Some(pre) =>
              JsonObj("@id" -> ("%s:%s" format (pre, suf)))
            case None =>
              JsonObj("@id" -> suf) } 

          addProps(obj, r, context, graph, query, prop2sn, drb, stack)

          obj }
      case r : Resource =>
        if(graph.listStatements(r, null, null : RDFNode).isEmpty && isObj) {
          if(drb.contains(r)) {
            JsonString("_:" + r.getId().getLabelString()) }
          else {
            JsonMap() }}
        else {
          val obj = if(drb.contains(r)) {
            JsonObj("@id" -> ("_:" + r.getId().getLabelString())) }
          else {
            JsonMap() }

          addProps(obj, r, context, graph, query, prop2sn, drb, stack)

          obj }
      case l : Literal =>
        if(l.getLanguage() != null && l.getLanguage() != "") {
          JsonObj("@value" -> l.getLexicalForm(),
                  "@language" -> l.getLanguage())  }
        else if(l.getDatatype() != null) {
          val (pre, suf) = splitURI(l.getDatatype().getURI())
          pre match {
            case Some(pre) =>
              JsonObj("@value" -> l.getLexicalForm(),
                      "@type" -> ("%s:%s" format (pre, suf)))
            case None =>
              JsonObj("@value" -> l.getLexicalForm(),
                      "@type" -> suf) }}
        else {
          return JsonString(l.getLexicalForm()) }}}

  private def doubleReffedBNodes(graph : Model) : Set[Resource] = {
    (for {
      o <- graph.listObjects()
      if o.isResource() && !o.isURIResource()
      if graph.listStatements(null, null, o).size > 1
    } yield o.asResource()).toSet }

  def write(out : java.io.Writer, graph : Model, query : String) {
    val obj = jsonLDfromModel(graph, query)
    obj.write(out, 0)
    out.write("\n")
    out.flush()
  }

  def jsonLDfromModel(graph : Model, query : String) : JsonObj = {
    val (context, prop2sn) = extractJsonLDContext(graph, query)
    val theId = if(query.startsWith(BASE_NAME)) {
        query.drop(BASE_NAME.size) }
      else {
        query }

    var theObj = JsonObj(
      "@context" -> context,
      "@id" -> JsonString(theId)
    )
    val elem = graph.createResource(query)

    val drb = doubleReffedBNodes(graph)
    addProps(theObj, elem, context, graph, query, prop2sn, drb, Nil)
    addInverseProps(theObj, elem, context, graph, query, prop2sn, drb, Nil)

    var rest = graph.listSubjects().toList
    if(!rest.isEmpty()) {
      val graphObj = JsonObj(
        "@context" -> context,
        "@graph" -> JsonList(Seq(theObj)))
      theObj.remove("@context")
      theObj = graphObj
      while(!rest.isEmpty()) {
        theObj("@graph") = theObj("@graph").asInstanceOf[JsonList] :+
          jsonLDValue(rest.head, context, graph, query, prop2sn, true, drb, Nil)
        rest = graph.listSubjects().toList }}

    theObj }
}

File: python/yuzu/jsonld.py
import re
from yuzu.settings import BASE_NAME
from rdflib.namespace import RDF
from rdflib import URIRef, BNode
import json


def is_alnum(string):
    return re.match('^\w+$', string) is not None


def prop_type(p, graph):
    if all(isinstance(o, URIRef) or isinstance(o, BNode)
           for o in graph.objects(None, p)):
        return {
            "@id": str(p),
            "@type": "@id"
        }
    else:
        return str(p)


def split_uri(value, graph):
    if value.startswith(BASE_NAME):
        return "", value[len(BASE_NAME):]
    for qn, uri in graph.namespaces():
        if value.startswith(uri):
            return qn, value[len(uri):]
    return "", value


def is_prefix_used(prefix, graph):
    for s, p, o in graph:
        if isinstance(s, URIRef) and str(s).startswith(prefix):
            return True
        if str(p).startswith(prefix):
            return True
        if isinstance(o, URIRef) and str(o).startswith(prefix):
            return True
    return False


def extract_jsonld_context(graph, query):
    context = {
        "@base": BASE_NAME}
    for k, v in graph.namespaces():
        if is_prefix_used(str(v), graph):
            context[k] = str(v)
    props = {}
    for p in set(graph.predicates()):
        p_str = str(p)
        short_name = ""
        if '#' in p_str:
            short_name = p_str[p_str.rindex('#')+1:]
        if not is_alnum(short_name) and '/' in p_str:
            short_name = p_str[p_str.rindex('/')+1:]
        if not is_alnum(short_name):
            pre, suf = split_uri(p_str, graph)
            short_name = suf
        sn = short_name
        i = 2
        while sn in context:
            sn = "%s%d" % (short_name, i)
            i += 1
        if p == RDF.type:
            sn = "@type"
        else:
            context[sn] = prop_type(p, graph)
        props[p_str] = sn
    return context, props


def add_props(obj, value, context, graph, query, prop2sn, drb, stack):
    if value not in stack:
        for p in set(graph.predicates(value)):
            objs = sorted(list(graph.objects(value, p)))
            is_obj = (p == RDF.type or
                      isinstance(context[prop2sn[str(p)]], dict))
            if len(objs) == 1:
                graph.remove((value, p, objs[0]))
                obj[prop2sn[str(p)]] = jsonld_value(objs[0], context, graph,
                                                    query, prop2sn, is_obj,
                                                    drb, [value] + stack)
            else:
                for o in objs:
                    graph.remove((value, p, o))
                obj[prop2sn[str(p)]] = [jsonld_value(o, context, graph, query,
                                                     prop2sn, is_obj, drb,
                                                     [value] + stack)
                                        for o in objs]


def add_inverse_props(obj, value, context, graph, query, prop2sn, drb, stack):
    for p in set(graph.predicates(value)):
        objs = sorted(list(graph.objects(value, p)))
        for o in objs:
            if isinstance(o, URIRef) or isinstance(o, BNode):
                add_inverse_props(obj[prop2sn[str(p)]], value, context, graph,
                                  query, prop2sn, drb, [value] + stack)

    if list(graph.predicates(None, value)):
        robj = {}
        obj["@reverse"] = robj
        for p in set(graph.predicates(None, value)):
            objs = sorted(list(graph.subjects(p, value)))
            #objs = [o for o in objs if not str(o).startswith(query)]
            if len(objs) == 1:
                graph.remove((objs[0], p, value))
                robj[prop2sn[str(p)]] = jsonld_value(objs[0], context, graph,
                                                     query, prop2sn, False,
                                                     drb, [value] + stack)
            elif len(objs) > 1:
                for o in objs:
                    graph.remove((o, p, value))
                robj[prop2sn[str(p)]] = [jsonld_value(o, context, graph, query,
                                                      prop2sn, False, drb,
                                                      [value] + stack)
                                         for o in objs]


def jsonld_value(value, context, graph, query, prop2sn, is_obj, drb, stack):
    if isinstance(value, list) and len(value) > 1:
        return [jsonld_value(v, context, graph, query, prop2sn,
                             is_obj, drb, stack)
                for v in value]
    elif isinstance(value, list):
        return jsonld_value(value[0], context, graph, query, prop2sn, is_obj,
                            drb, stack)
    elif isinstance(value, URIRef):
        if not list(graph.predicate_objects(value)) and is_obj:
            pre, suf = split_uri(str(value), graph)
            if pre:
                return "%s:%s" % (pre, suf)
            else:
                return suf
        else:
            pre, suf = split_uri(str(value), graph)
            if pre:
                obj = {"@id": "%s:%s" % (pre, suf)}
            else:
                obj = {"@id": suf}

            add_props(obj, value, context, graph, query, prop2sn, drb,
                      stack)

            return obj
    elif isinstance(value, BNode):
        if not list(graph.predicate_objects(value)) and is_obj:
            if value in drb:
                return "_:" + str(value)
            else:
                return {}
        else:
            if value in drb:
                obj = {"@id": "_:" + str(value)}
            else:
                obj = {}

            add_props(obj, value, context, graph, query, prop2sn, drb,
                      stack)

            return obj
    else:
        if value.language:
            return {"@value": str(value),
                    "@language": value.language}
        elif value.datatype:
            pre, suf = split_uri(value.datatype, graph)
            if pre == "xsd" and suf == "integer":
                return int(value)
            elif pre == "xsd" and suf == "double":
                return float(value)
            elif pre:
                return {"@value": str(value),
                        "@type": str("%s:%s" % (pre, suf))}
            else:
                return {"@value": str(value),
                        "@type": str(suf)}
        else:
            return str(value)


def double_reffed_bnodes(graph):
    for o in graph.objects():
        if isinstance(o, BNode):
            if len(list(graph.subject_predicates(o))) > 1:
                yield(o)


def write(graph, query):
    return json.dumps(jsonld_from_model(graph, query), indent=2)


def jsonld_from_model(graph, query):
    context, prop2sn = extract_jsonld_context(graph, query)
    if query.startswith(BASE_NAME):
        the_id = query[len(BASE_NAME):]
    else:
        the_id = query
    the_obj = {
        "@context": context,
        "@id": the_id
    }
    elem = URIRef(query)

    drb = list(double_reffed_bnodes(graph))
    add_props(the_obj, elem, context, graph, query, prop2sn, drb, [])
    add_inverse_props(the_obj, elem, context, graph, query, prop2sn, drb, [])

    rest = list(graph.subjects())
    if rest:
        graph_obj = {
            "@context": context,
            "@graph": [
                the_obj
            ]
        }
        del the_obj["@context"]
        the_obj = graph_obj
        while rest:
            the_obj["@graph"].append(jsonld_value(rest[0], context, graph,
                                                  query, prop2sn, True, drb,
                                                  []))
            rest = list(graph.subjects())

    return the_obj
